File: solutions/python/trees/maximum-binary-tree.py
# LeetCode 654 — Maximum Binary Tree (Medium)
# Category: Trees · Approach: Recursive max-split
# Time: O(n²) | Space: O(n)
# Source: https://leetcode.com/problems/maximum-binary-tree/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val, self.left, self.right = val, left, right

def construct_maximum_binary_tree(nums: list[int]) -> TreeNode | None:
    def build(lo: int, hi: int) -> TreeNode | None:
        if lo > hi:
            return None
        mi = lo
        for i in range(lo + 1, hi + 1):
            if nums[i] > nums[mi]:
                mi = i
        return TreeNode(nums[mi], build(lo, mi - 1), build(mi + 1, hi))
    return build(0, len(nums) - 1)

File: solutions/typescript/trees/maximum-binary-tree.ts
// LeetCode 654 — Maximum Binary Tree (Medium)
// Category: Trees · Approach: Recursive max-split
// Time: O(n²) | Space: O(n)
// Source: https://leetcode.com/problems/maximum-binary-tree/

class TreeNode {
  val: number;
  left: TreeNode | null = null;
  right: TreeNode | null = null;
  constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function constructMaximumBinaryTree(nums: number[]): TreeNode | null {
  function build(lo: number, hi: number): TreeNode | null {
    if (lo > hi) return null;
    let mi = lo;
    for (let i = lo + 1; i <= hi; i++) if (nums[i] > nums[mi]) mi = i;
    return new TreeNode(nums[mi], build(lo, mi - 1), build(mi + 1, hi));
  }
  return build(0, nums.length - 1);
}
